### PR TITLE
Add metadata validation rule: either and precludes

### PR DESCRIPTION
**Pull request by LowEntropyBody**

### CODE REVIEW

**Comment by yufeih** [comment]
@LowEntropyBody , could you link to the original issue?

File: src/docfx/lib/schema/JsonSchemaValidator.cs [hunk 1]
```diff
@@ -155,6 +155,22 @@ private void ValidateObject(JsonSchema schema, List<Error> errors, JObject map)
                 }
             }
 
+            foreach (var keys in schema.Either)
```
Comment (yufeih) [inline src/docfx/lib/schema/JsonSchemaValidator.cs]: Can you help do a simple refactor to extract these blocks to `ValidateEither`, `ValidateAdditionalProperties`, etc...

Comment (LowEntropyBody) [reply src/docfx/lib/schema/JsonSchemaValidator.cs]: Sure, I will do this

**Comment by LowEntropyBody** [comment]
Related issue: #4605

File: src/docfx/lib/schema/JsonSchema.cs [hunk 2]
```diff
@@ -116,5 +116,18 @@ internal class JsonSchema
...
+        // JSON schema metadata validation extensions
+        //-------------------------------------------
+
+        /// <summary>
+        /// Properties that are used to realize either logic
+        /// </summary>
+        public List<string[]> Either { get; set; } = new List<string[]>();
+
+        /// <summary>
+        /// Properties that are are used to realize precludes logic
```
Comment (mingwli) [inline src/docfx/lib/schema/JsonSchema.cs]: > are  [](start = 31, length = 5)

typo

File: src/docfx/lib/schema/JsonSchemaValidator.cs [hunk 3]
```diff
@@ -132,34 +149,75 @@ private void ValidateObject(JsonSchema schema, List<Error> errors, JObject map)
...
+        {
+            int existNum;
+            foreach (var keys in schema.Precludes)
             {
-                if (schema.Properties.TryGetValue(key, out var propertySchema))
+                existNum = 0;
+                foreach (var key in keys)
                 {
-                    Validate(propertySchema, value, errors);
+                    if (map.ContainsKey(key))
```
Comment (mingwli) [inline src/docfx/lib/schema/JsonSchemaValidator.cs]: >if (map.ContainsKey(key)) [](start = 20, length = 25)

if (map.ContainsKey(key) && ++existNum > 1) // add error and break earlier?

File: src/docfx/lib/schema/JsonSchema.cs [hunk 4]
```diff
@@ -116,5 +116,18 @@ internal class JsonSchema
...
         /// </summary>
         public string[] XrefProperties { get; set; } = Array.Empty<string>();
+
+        // JSON schema metadata validation extensions
+        //-------------------------------------------
+
+        /// <summary>
+        /// Properties that are used to realize either logic
+        /// </summary>
+        public List<string[]> Either { get; set; } = new List<string[]>();
```
Comment (yufeih) [inline src/docfx/lib/schema/JsonSchema.cs]:
```suggestion
        public string[][] Either { get; set; } = Array.Empty<string[]>();
```

same as below

File: src/docfx/lib/schema/JsonSchemaValidator.cs [hunk 5]
```diff
@@ -132,34 +149,75 @@ private void ValidateObject(JsonSchema schema, List<Error> errors, JObject map)
...
+        }
 
-            foreach (var (key, value) in map)
+        private void ValidatePrecludes(JsonSchema schema, List<Error> errors, JObject map)
+        {
+            int existNum;
+            foreach (var keys in schema.Precludes)
             {
-                if (schema.Properties.TryGetValue(key, out var propertySchema))
+                existNum = 0;
```
Comment (yufeih) [inline src/docfx/lib/schema/JsonSchemaValidator.cs]: nit: 
```suggestion
                var existNum = 0;
```

File: src/docfx/lib/schema/JsonSchemaValidator.cs [hunk 6]
```diff
@@ -132,34 +149,75 @@ private void ValidateObject(JsonSchema schema, List<Error> errors, JObject map)
...
+        }
 
-            foreach (var (key, value) in schema.Dependencies)
+        private void ValidateEither(JsonSchema schema, List<Error> errors, JObject map)
+        {
+            bool result;
+            foreach (var keys in schema.Either)
             {
-                if (map.ContainsKey(key))
+                result = false;
```
Comment (yufeih) [inline src/docfx/lib/schema/JsonSchemaValidator.cs]: nit:
```suggestion
                var result = false;
```

**Comment by LowEntropyBody** [comment]
Work item link: [100602](https://dev.azure.com/ceapex/Engineering/_workitems/edit/100602)

**Pull request merged**